### DISCUSSION TITLE
Support dict type for media_player source_list and sound_mode_list

### DISCUSSION THREAD

**Discussion by micha91** [home-assistant/architecture#641] [Entity Models] [answered]
# Context
First of all two definitions to prevent confusion:
* `source-name` the actual name of a source (normally defined by the manufacturer of the device)
* `source-label` a more user friendly definition of a source (often configurable by the user)

Because these `source-labels` are more user friendly, many media_players use them to display the sources in UI. At the moment they have to add these values to the `source_list` and do a reverse mapping from these values to the actual source name when the `select_source` service is called. Here some examples for integrations going this way:
* yamaha
* onkyo
* nad
* denon

In addition about 30 additional integrations are facing this challenge and solve it in a similar way (see [here](https://github.com/home-assistant/architecture/discussions/641#discussioncomment-1676896)).

Some of these integrations take these `source-labels` from the device itself (for Yamaha devices they can be defined in their app for example), some are using a fixed list in the code and some are using the configuration.yaml. The main problem with this is, that it is not guaranteed that these user defined labels are unique. On Yamaha devices for example, it is possible to define Apple-TV as label for multiple different inputs. It's also possible to set the label of a source to the name of another input. For example I had an echo dot connected to aux1 of my AV-Receiver and set the label of this input to Alexa. With a new firmware update Yamaha added a new source named Alexa.

We have the same challenge for sound modes, where an id or name must be sent to the device, but there maybe is a more user friendly representation of that, which should be displayed in the UI.
# Proposal
We allow `source_list` and `sound_mode_list` to be dictionaries. In the frontend, we will detect, whether they are lists or dictionaries and built up the option picker accordingly. If these properties are dictionaries, the key will be the `name` and the value the `label`.

**Comment by micha91** [comment]
Because of the already existing possibility to customize the state attributes, I would still go this way. In my opinion, we should add labels for sources as well as for sound modes to the media player and for options to the select entity. 
It could also be an option to let source_list, sound_mode_list and (for select) options become dictionaries (while still supporting lists of course). This way we would not need additional state attributes and the frontend could check, if the list of options is a dictionary and select the right way to display it.

**Reply by micha91** [reply]
I just took a deeper look into the implementations of the select_source service of all integrations. The majority does use a `source_label` for the implementation of the service and does some reverse mapping afterwards, because some kind of `source_name` needs to be sent to the device. So this is a topic, which is related to a huge amount of integrations. Here on overview of my results:
- PS4 (uses select source to select a game and has to browse all games titles to find the correct ID)
- Vizio (uses select source to start apps. App names are used to find the next matching app config and this config is used to start the app. After some refactoring this could also use app ids and names)
- Songpal (Uses an dictionary with URLs as keys and special input objects as values. These input objects are browsed for their title and the URL is used to set the source)
- Roku (Has app ID and app name in a special app object, which are both allowed to select the source, uses app name for source list)
- Epson (Has a DEFAULT_SOURCES dictionary with labels and IDs for all sources in the background)
- WebOSTV (has a source dictionary with IDs and label/titles. Labels and titles are used to call the HA service, while int the background the ID is needed.)
- BraviaTV (uses URIs to select a source. The title is used as key in dictionary and the URI as value and must be sent to the device)
- Denonavr (uses a mapping from source labels to the source IDs in the library)
- Openhome (uses a dictionary from source name to source index, because the source index needs to be send to their library)
- Andoidtv (uses select_source to start/stop an app. Apps have IDs and they names are mapped to their IDs in a separate dictionary)
- arcam_fmj (uses an enum for its source names to ID mapping)
- Bluesound (uses the title of its presets as well as service and capture items and selects the first matching item. In the end the URL of the item is always needed to be send to the device.)
- philips_js (has a sources dictionary with the ID as key and the title as value. The inverted dict is used to find the correct source_id)
- forked_daapd (needs to use URIs to select a source. Uses a title -> URI mapping to do so)
- Russound_rio (uses a dictionary of source names to source_ids, which need to be send to the device)
- homekit_controller (uses a special services structure and filters it for the input with the requested name. In the end the identifier of this service is needed to change the source)
- nad (uses a reverse mapping from source titles to IDs)
- denon (uses a dictionary for the mapping of input names to ids)
- onkyo (uses a dictionary for the mapping of input names to ids, adjustable in the config.yaml)
- Pijlink (uses a dictionary for the mapping of input names to ids)
- Yamaha (uses a dictionary for the mapping of input names to ids, adjustable in the config.yaml)
- aquostv (Has a fixed dictionary from source keys to source names)
- Enigma2 (Has a sources dictionary in the library)
- pandora (has a source list and needs to get the index from it to send it to the device)
- pioneer (uses a source name to number mapping)
- channels (Has a list of favorite channels. Their names are used to be displayed, but the number needs to be send to the device)
- blackbird (Has a source name -> ID mapping, as the ID needs to be sent to the device, configurable in the config.yaml)
- monoprice (Has multiple mappings for source names/IDs and supports a configuration using config.yaml)
- samsungtv (Has a fixed mapping from source names to source keys)
- clementine (Uses select source to select playlists. The playlists are displayed using their name, but the ID needs to be sent to the device. For that reason the list of playlists is browsed for the correct playlist)
- lg_netcast (Uses the select source service to select channels. The channels  are displayed using their names and need to be selected using their numbers)
- russound_rnet (Uses the index of the source in the list of sources. Source names are configurable in the config.yaml)
- ziggo_mediabox_xl (Uses the select source service to set channels. The channels need to be set by their number)

All these integrations could potentially benefit of the proposed change. Of course not all of these integration have a high risk of male functioning due to the fact that we do not make a difference between the displayed and the requested source values. But in my opinion, it would bring us many benefits, if we would add a `source-name`-`source-label` mapping or allow the source_list to be a dictionary (as an alternative to the current list type):
- Users of all media players could change the labels of sources in the customize.yaml (Currently some integrations support this using the config.yaml)
- We would ensure that duplicate labels/titles of sources do not cause male functionality
- Automations could always call the select_source service with reliable `source_names`, which cannot be changed by the user

**Comment by elupus** [comment]
Select entity have the same needs. If the list of sources would just return a dict with key, title the UI could use that.

**Reply by micha91** [reply]
Yes, I think so, too. `source_list` and `sound_mode_list` for media players and for select entities `options` could simply add dict as a valid type. In the frontend the key should be used to call the service while the value is used to display the entry. If the attribute is a list, the behavior can be kept as it is now. Using the customize.yaml users could also add their own titles for options, so that these could also be language specific by manually setting them, if the user wants that.
Do you think that it makes sense to rewrite the initial post to cover the whole topic?

**Reply by elupus** [reply]
I think we can keep this topic on media player source to not expand it too much. But keep in mind that there are other places same thing is needed.

**Reply by micha91** [reply]
I just updated the title and proposal of this discussion to match the current approach of adding an additional type for source_list and sound_mode_list

**Comment by balloob** [comment]
The problem with defining titles in the backend is that you can't translate them.

**Reply by micha91** [reply]
Yes, that's true. But especially for sources, I believe there is often no specific title per language. For me hdmi1 is "Apple-TV", but for my neighbor it might be "XBOX" while someone in Spain wants to call it "Apple-TV", too. If we support the dict type for source_list, integrations, which get the title from the device, could fill this dictionary. If someone wants to change this in his installation, he could still change the dictionary in the customize.yaml. In addition, we get the available sources in many cases from the device, so that we may not even know all allowed values to place them in a translation file.

For the sound_mode_list translations could be more important. My device has for example sound modes named "Role-playing Game", "Sports", "Adventure" and others. Having a translation for those, would totally make sense to me, as their title is not user but language specific. We still have the problem here that we do not know all available sound modes across all devices, but we can collect them over time with the help of the users, so this is not a big deal for me.

**Comment by tomlut** [comment]
Did this get anywhere?

It's the only thing stopping me moving from the Yamaha Network Receivers integration to the MusicCast Integration (which has a lot of advantages). 

The Network Receivers integration allows hiding and relabelling sources:

```
media_player:
  - platform: yamaha
    host: 192.168.0.10
    source_ignore:
      - "AUX"
      - "HDMI6"
    source_names:
      HDMI1: "ChromeCast"
      AV4: "Vinyl"
```

https://www.home-assistant.io/integrations/yamaha/#example-configuration

**Reply by mvdwetering** [reply]
Hiding sources and relabeling names are just features that are not implemented in the MusicCast integration. Implementing those features should not need "dict support for sources/soundmodes" as described in this discussion.

**Reply by micha91** [reply]
First of all, hiding sources is possible using the customize.yaml for any media_player - this is not a platform specific thing and does not need any adjustments in the MusicCast integration at all. However this discussion is not about the MusicCast integration, it's about the general problem of shown labels vs. names for sources and sound modes. I will try to explain the reason for this discussion again using the MusicCast integration as an example. 
As I mentioned [here](https://github.com/home-assistant/core/issues/56525#issuecomment-926811157) there are many edge cases, which would not be covered by using the source names, which were configured by the user in the MusicCast App, as the source names in HA. It is for example possible to label hdmi1 and hdmi2 both with Apple-TV in the MusicCast App. If we would now not display hdmi1 and hdmi2 but Apple-TV twice to the frontend, the select_source service would be called with "Apple-TV" whenever you select the hdmi1 or hdmi2 source - how should the backend now know, whether it should switch to hdmi1 or hdmi2? Obviously we could say it's the users fault and we always switch to hdmi1, but I think, it would be much cleaner to display the user defined label (such as Apple-TV) and call the service with a unique name (such as hdmi1). If it comes to sound modes, this is even worse. There we have sound modes like "action_game", "munich", or "bottom_line". These should maybe even be translated depending on the user interface language.
However, obviously finding a general solution for this issue is more a marathon than a short run. I will offer an intermediate solution for MusicCast source names and sound modes soon, but I still hope that we will find a general clean solution for this some day in this discussion.

**Comment by frenck** [comment, marked as answer]
This proposal is no longer needed, as state attribute values can now be translated.

This already works for, for example, the automation editors. For the more information dialogs, this is being fixed/added in: <https://github.com/home-assistant/frontend/pull/16559>

../Frenck
